### PR TITLE
List: items with index bigger then 255 can be selected in grouped list (T996851)

### PR DESCRIPTION
**Pull request by Zedwag**

### CODE REVIEW

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js [hunk 1]
```diff
@@ -125,6 +125,33 @@ QUnit.test('index should be correct for grouped list', function(assert) {
...
+    const items = [];
+    for(let i = 0; i < 300; i += 1) {
+        items.push(i);
+    }
+    const dataSource = [{
+        key: 'someKey',
+        items
+    }];
+    let selectedIdx;
+    const list = $('#templated-list').dxList({
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: should we use exactly `templated-list`, not just `list`?

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js [hunk 2]
```diff
@@ -125,6 +125,33 @@ QUnit.test('index should be correct for grouped list', function(assert) {
...
+QUnit.test('grouped list can select items with index more then 255 (T996851)', function(assert) {
+    const items = [];
+    for(let i = 0; i < 300; i += 1) {
+        items.push(i);
+    }
+    const dataSource = [{
+        key: 'someKey',
+        items
+    }];
+    let selectedIdx;
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: let's do not use not generally accepted abbreviated words
```suggestion
    let selectedIndex;
```

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js [hunk 3]
```diff
@@ -125,6 +125,33 @@ QUnit.test('index should be correct for grouped list', function(assert) {
...
+    const list = $('#templated-list').dxList({
+        dataSource,
+        grouped: true,
+        selectionMode: 'single',
+        onSelectionChanged: function(e) {
+            selectedIdx = e.addedItems[0].text;
+        },
+        groupTemplate: function(data) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: seems like excess data passing

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js (hunk 2 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: done

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js (hunk 3 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: done

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js (hunk 1 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: changed

File: js/ui/list/ui.list.edit.strategy.grouped.js [hunk 4]
```diff
@@ -19,7 +18,7 @@ const combineIndex = function(indices) {
 const splitIndex = function(combinedIndex) {
     return {
         group: combinedIndex >> SELECTION_SHIFT,
-        item: combinedIndex & SELECTION_MASK
+        item: combinedIndex & ((1 << SELECTION_SHIFT) - 1)
```
Comment (ksercs) [inline js/ui/list/ui.list.edit.strategy.grouped.js]: It's good idea to specify a constant to not calculate this value each time

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js [hunk 5]
```diff
@@ -125,6 +125,33 @@ QUnit.test('index should be correct for grouped list', function(assert) {
...
+    }];
+    let selectedIndex;
+    const list = $('#list').dxList({
+        dataSource,
+        grouped: true,
+        selectionMode: 'single',
+        onSelectionChanged: function(e) {
+            selectedIndex = e.addedItems[0].text;
+        },
+        groupTemplate: function() {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: do we really need specify `groupTemplate`?

File: js/ui/list/ui.list.edit.strategy.grouped.js (hunk 4 above)
Comment (Zedwag) [reply js/ui/list/ui.list.edit.strategy.grouped.js]: Fair enough

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js [hunk 6]
```diff
@@ -125,6 +125,33 @@ QUnit.test('index should be correct for grouped list', function(assert) {
     assert.equal(list.getItemElementByFlatIndex(5).get(0), $item.get(0), 'item correct');
 });
 
+QUnit.test('it should be possible to select an item with index bigger then 255 in the grouped list (T996851)', function(assert) {
+    const items = [];
+    for(let i = 0; i < 300; i += 1) {
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: we can write just
`[...Array(300).keys()]`

**Review by ksercs — DISMISSED** [review]

File: js/ui/list/ui.list.edit.strategy.grouped.js [hunk 7]
```diff
@@ -10,7 +10,7 @@ const LIST_ITEM_CLASS = 'dx-list-item';
 const LIST_GROUP_CLASS = 'dx-list-group';
 
 const SELECTION_SHIFT = 20;
-const SELECTION_MASK = 0x8FF;
+const SELECTION_MASK = ((1 << SELECTION_SHIFT) - 1);
```
Comment (ksercs) [inline js/ui/list/ui.list.edit.strategy.grouped.js]: we don't need outer brackets anymore

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js [hunk 8]
```diff
@@ -125,6 +125,33 @@ QUnit.test('index should be correct for grouped list', function(assert) {
...
+        items.push(i);
+    }
+    const dataSource = [{
+        key: 'someKey',
```
Comment (ksercs) [inline testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: do we need specify `key`? Do we need specify exactly `dataSource`, not just pass `items`?

Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: https://js.devexpress.com/Documentation/ApiReference/UI_Components/dxList/Configuration/#grouped

File: js/ui/list/ui.list.edit.strategy.grouped.js (hunk 7 above)
Comment (Zedwag) [reply js/ui/list/ui.list.edit.strategy.grouped.js]: done

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js (hunk 6 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: yep

File: testing/tests/DevExpress.ui.widgets/listParts/editingTests.js (hunk 5 above)
Comment (Zedwag) [reply testing/tests/DevExpress.ui.widgets/listParts/editingTests.js]: nope

**Review by ksercs — APPROVED** [review]

**Review by alexander-kotov-dx — APPROVED** [review]

**Review by DokaRus — APPROVED** [review]
nice

**Pull request merged**